File: haskell/src/Solutions/Euler358.hs
module Solutions.Euler358 where
import Data.List(find)
import qualified Data.Map.Strict as M
import Utils.NumberTheory
import Utils.Numeric(fastPowerMod, digitsSum)
import Utils.Operators

f = fastPowerMod 10
fromJust (Just x) = x

primRoot :: Prime Int -> Bool
primRoot n' = all (\p -> f (phi `div` unPrime p) n /= 1) ps 
    where n = unPrime n'
          phi = n - 1
          ps = M.keys $ runFactorization $ factorize phi

properEnd :: Prime Int -> Bool
properEnd n' = ((f (n - 1) d - 1) * fromJust (invertMod n d)) `mod` d == 56789
    where n = unPrime n'
          d = 100000

uniquePrime :: Int
uniquePrime = unPrime $ fromJust $ find
                 (primRoot <&&> properEnd)
                 [nextPrime 724637680 .. precPrime 729927080]
                 -- these bounds are for 00000000137 part, found by plugging 10^11/137 and 10^11/138 to WolframAlpha

cyclicDigitsSum :: Int -> Int
cyclicDigitsSum n = go 0 9 where
    go acc rem
        | r == 0 = acc + q
        | otherwise = go (acc + q) (10 * r + 9) 
        where (q, r) = rem `quotRem` n

euler358 :: IO String
euler358 = return $ show $ cyclicDigitsSum uniquePrime

File: rust/src/solutions/euler358.rs
use crate::utils::number_theory::invert_mod::InvertMod;
use crate::utils::numeric::{factorization::Factorization, modpow::ModPow};
use primal::Sieve;

fn f(e: i64, m: i64) -> i64 {
    10.modpow(&e, &m)
}

fn prim_root(sieve: &Sieve, n: i64) -> bool {
    let phi = n - 1;
    Factorization::new(phi as usize, sieve)
        .factors()
        .into_iter()
        .all(|(p, _)| f(phi / p as i64, n) != 1)
}

fn proper_end(n: i64) -> bool {
    let d = 100_000;
    let g = |x| {
        let y = x % d;
        if y < 0 {
            y + d
        } else {
            y
        }
    };
    g((f(n - 1, d) - 1) * n.invert_mod(&d).unwrap()) == 56789
}

fn cyclic_digits_sum(n: i64) -> i64 {
    let mut acc = 0;
    let mut rem = 9;
    loop {
        acc += rem / n;
        let r = rem % n;
        if r == 0 {
            break acc;
        } else {
            rem = 10 * r + 9;
        }
    }
}

pub fn euler358() -> String {
    let sieve = Sieve::new(100_000);
    // bounds are for 00000000137 part, found by plugging 10^11/137 and 10^11/138 to WolframAlpha
    let unique_prime = (724_637_680..=729_927_080)
        .filter(|&p| primal::is_prime(p))
        .map(|p| p as i64)
        .find(|&p| prim_root(&sieve, p) && proper_end(p))
        .unwrap();
    cyclic_digits_sum(unique_prime).to_string()
}
